File: apps/token-app-api/constants/swaggerOptions/paths/bankCard.js
module.exports = {
  '/bankcards/{bankcard_id}': {
    put: {
      tags: ['銀行卡'],
      summary: '編輯銀行卡',
      description: '編輯銀行卡',
      operationId: 'updatebandcard',
      consumes: ['application/json'],
      produces: ['application/json'],
      parameters: [
        {
          in: "header",
          name: "Authorization",
          description: "Bearer ${{token}}",
          required: true,
          schema: {
            type: "string",
          },
        },
        {
          in: "path",
          name: "bankcard_id",
          description: "銀行卡ID",
          required: true,
          schema: {
            type: "string",
          },
          default: 1,
        },
        {
          in: "body",
          type: "object",
          name: "data",
          required: true,
          schema: {
            type: "object",
            $ref: "#/definitions/UpdateBankcard",
          },
        }
      ],
      responses: {
        200: {
          description: '新增成功',
          schema: {
            type: "object",
            properties: {
              success: {
                type: "boolean",
                default: true,
              },
              data: {
                type: "object",
                $ref: "#/definitions/Bankcard",
              },
            },
          },
        },
      },
    },
    delete: {
      tags: ['銀行卡'],
      summary: '刪除銀行卡',
      description: '刪除銀行卡',
      operationId: 'deletebandcard',
      consumes: ['application/json'],
      produces: ['application/json'],
      parameters: [
        {
          in: "header",
          name: "Authorization",
          description: "Bearer ${{token}}",
          required: true,
          schema: {
            type: "string",
          },
        },
        {
          in: "path",
          name: "bankcard_id",
          description: "銀行卡ID",
          required: true,
          schema: {
            type: "string",
          },
          default: 1,
        },
      ],
      responses: {
        200: {
          description: '刪除成功',
          schema: {
            type: "object",
            properties: {
              success: {
                type: "boolean",
                default: true,
              },
            },
          },
        },
      },
    },
  },
  '/bankcards': {
    get: {
      tags: ['銀行卡'],
      summary: '取回銀行卡列表',
      description: '取回銀行卡列表',
      operationId: 'getbandcard',
      consumes: ['application/json'],
      produces: ['application/json'],
      parameters: [
        {
          in: "header",
          name: "Authorization",
          description: "Bearer ${{token}}",
          required: true,
          schema: {
            type: "string",
          },
        },
      ],
      responses: {
        200: {
          description: '取回列表成功',
          schema: {
            type: "object",
            properties: {
              success: {
                type: "boolean",
                default: true,
              },
              data: {
                type: "array",
                $ref: "#/definitions/Bankcards",
              },
            },
          },
        },
      },
    },
    post: {
      tags: ['銀行卡'],
      summary: '新增銀行卡',
      description: '新增銀行卡',
      operationId: 'createbandcard',
      consumes: ['application/json'],
      produces: ['application/json'],
      parameters: [
        {
          in: "header",
          name: "Authorization",
          description: "Bearer ${{token}}",
          required: true,
          schema: {
            type: "string",
          },
        },
        {
          in: "body",
          type: "object",
          name: "data",
          required: true,
          schema: {
            type: "object",
            $ref: "#/definitions/CreateBankcard",
          },
        }
      ],
      responses: {
        200: {
          description: '新增成功',
          schema: {
            type: "object",
            properties: {
              success: {
                type: "boolean",
                default: true,
              },
              data: {
                type: "object",
                $ref: "#/definitions/Bankcard",
              },
            },
          },
        },
      },
    },
  },
};

File: cmd/token-app-api/internal/handlers/bankcard.go
package handlers

import (
	"strconv"

	"passontw-backend-services/cmd/token-app-api/internal/interfaces"
	"passontw-backend-services/pkg/logger"
	"passontw-backend-services/pkg/response"

	"github.com/gin-gonic/gin"
)

type BankCardHandlers struct {
	bankCardService interfaces.BankCardServiceInterface
	logger          logger.Logger
}

func NewBankCardHandlers(bankCardService interfaces.BankCardServiceInterface, logger logger.Logger) *BankCardHandlers {
	return &BankCardHandlers{
		bankCardService: bankCardService,
		logger:          logger,
	}
}

// ==================== BankCard Handlers ====================

// GetBankCards godoc
// @Summary 取回銀行卡列表
// @Description 取回使用者的銀行卡列表
// @Tags 銀行卡
// @Accept json
// @Produce json
// @Security Bearer
// @Success 200 {object} response.Response{data=[]interfaces.BankCardDetail}
// @Failure 401 {object} response.ErrorResponse
// @Failure 500 {object} response.ErrorResponse
// @Router /bankcards [get]
func (h *BankCardHandlers) GetBankCards(c *gin.Context) {
	userID := c.GetInt("user_id")

	bankCards, err := h.bankCardService.GetBankCards(userID)
	if err != nil {
		response.InternalError(c)
		return
	}

	response.SuccessWithMessage(c, bankCards)
}

// CreateBankCard godoc
// @Summary 新增銀行卡
// @Description 新增銀行卡
// @Tags 銀行卡
// @Accept json
// @Produce json
// @Security Bearer
// @Param data body interfaces.CreateBankCardRequest true "銀行卡資訊"
// @Success 200 {object} response.Response{data=interfaces.BankCardDetail}
// @Failure 400 {object} response.ErrorResponse
// @Failure 401 {object} response.ErrorResponse
// @Router /bankcards [post]
func (h *BankCardHandlers) CreateBankCard(c *gin.Context) {
	userID := c.GetInt("user_id")

	var req interfaces.CreateBankCardRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		response.BadRequest(c)
		return
	}

	bankCard, err := h.bankCardService.CreateBankCard(userID, &req)
	if err != nil {
		response.BadRequest(c)
		return
	}

	response.SuccessWithMessage(c, bankCard)
}

// UpdateBankCard godoc
// @Summary 編輯銀行卡
// @Description 編輯銀行卡資訊
// @Tags 銀行卡
// @Accept json
// @Produce json
// @Security Bearer
// @Param bankcard_id path int true "銀行卡 ID"
// @Param data body interfaces.UpdateBankCardRequest true "更新資訊"
// @Success 200 {object} response.Response{data=interfaces.BankCardDetail}
// @Failure 400 {object} response.ErrorResponse
// @Failure 401 {object} response.ErrorResponse
// @Failure 404 {object} response.ErrorResponse
// @Router /bankcards/{bankcard_id} [put]
func (h *BankCardHandlers) UpdateBankCard(c *gin.Context) {
	userID := c.GetInt("user_id")

	bankcardID, err := strconv.Atoi(c.Param("bankcard_id"))
	if err != nil {
		response.BadRequest(c)
		return
	}

	var req interfaces.UpdateBankCardRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		response.BadRequest(c)
		return
	}

	bankCard, err := h.bankCardService.UpdateBankCard(userID, bankcardID, &req)
	if err != nil {
		response.BadRequest(c)
		return
	}

	response.SuccessWithMessage(c, bankCard)
}

// DeleteBankCard godoc
// @Summary 刪除銀行卡
// @Description 刪除銀行卡
// @Tags 銀行卡
// @Accept json
// @Produce json
// @Security Bearer
// @Param bankcard_id path int true "銀行卡 ID"
// @Success 200 {object} response.Response
// @Failure 400 {object} response.ErrorResponse
// @Failure 401 {object} response.ErrorResponse
// @Failure 404 {object} response.ErrorResponse
// @Router /bankcards/{bankcard_id} [delete]
func (h *BankCardHandlers) DeleteBankCard(c *gin.Context) {
	userID := c.GetInt("user_id")

	bankcardID, err := strconv.Atoi(c.Param("bankcard_id"))
	if err != nil {
		response.BadRequest(c)
		return
	}

	if err := h.bankCardService.DeleteBankCard(userID, bankcardID); err != nil {
		response.BadRequest(c)
		return
	}

	response.SuccessWithMessage(c, nil)
}
